src/v/config/tests/enterprise_property_test.cc

```diff
@@ -96,6 +100,108 @@ TEST(EnterprisePropertyTest, TestRestriction) {
     EXPECT_TRUE(cfg.enterprise_enum.check_restricted(N(tls_version::v1_3)));
 }
 
+template<typename Property>
+struct EnterprisePropertyTest : public ::testing::Test {
+    using value_type = typename Property::underlying;
+
+    struct test_values {
+        value_type default_value;
+        value_type allowed_value;
+        value_type restricted_value;
+    };
+
+    struct test_case {
+        enterprise<Property>* enterprise_conf;
+        test_values ts;
+    };
+
+    test_case get_test_case() {
+        if constexpr (std::same_as<Property, property<bool>>) {
+            return {
+              &cfg.enterprise_bool,
+              {.default_value = false,
+               .allowed_value = false,
+               .restricted_value = true}};
+        }
+        if constexpr (std::same_as<Property, enum_property<ss::sstring>>) {
+            return {
+              &cfg.enterprise_str_enum,
+              {.default_value = "foo",
+               .allowed_value = "baz",
+               .restricted_value = "bar"}};
+        }
+        if constexpr (std::
+                        same_as<Property, property<std::vector<ss::sstring>>>) {
+            return {
+              &cfg.enterprise_str_vec,
+              {.default_value{},
+               .allowed_value = {"OTHER"},
+               .restricted_value = {"GSSAPI", "OTHER"}}};
+        }
+        if constexpr (std::same_as<Property, property<std::optional<int>>>) {
+            return {
+              &cfg.enterprise_opt_int,
+              {.default_value{0}, .allowed_value{10}, .restricted_value{1010}}};
+        }
+        if constexpr (std::same_as<Property, enum_property<tls_version>>) {
+            return {
+              &cfg.enterprise_enum,
+              {.default_value = tls_version::v1_1,
+               .allowed_value = tls_version::v1_2,
+               .restricted_value = tls_version::v1_3}};
+        }
+    }
+
+    test_config cfg{};
+};
+
+using EnterprisePropertyTypes = ::testing::Types<
+  property<bool>,
+  enum_property<ss::sstring>,
+  property<std::vector<ss::sstring>>,
+  property<std::optional<int>>,
+  enum_property<tls_version>>;
+TYPED_TEST_SUITE(EnterprisePropertyTest, EnterprisePropertyTypes);
+
+TYPED_TEST(EnterprisePropertyTest, SanctionedValues) {
```

Comment:
thought: since the new API always returns the value type of the underlying property, do we gain much coverage by testing every last type? maybe all we need to test is that `enterprise::maybe_sanctioned_value` returns the default value if sanctioned, configured value otherwise.

Reply:
I wrote once for each for sanity checks that it works with various types. And the options you picked for the testing offer a very nice spread.

However, I do agree that they are very verbose. Happy to dial back to just one type, if you think it is too much.

Reply:
Eh, more tests probably better. Just a thought!